hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsV2ResourceCalculator.java

```diff
@@ -0,0 +1,135 @@
+ * /proc/{pid}/cgroup file currently we can not handle multiple lines.
+ */
+public class CGroupsV2ResourceCalculator extends AbstractCGroupsResourceCalculator {
+  private static final Logger LOG = LoggerFactory.getLogger(CGroupsV2ResourceCalculator.class);
+  private final Map<String, String> stats = new ConcurrentHashMap<>();
+
+  @VisibleForTesting
+  String root = "/";
+
+  /**
+   * Create resource calculator for the container that has the specified pid.
+   * @param pid A pid from the cgroup or null for all containers
+   */
+  public CGroupsV2ResourceCalculator(String pid) {
+    super(pid);
+  }
+
+  @Override
+  public void initialize() throws YarnException {
+    cpuTimeTracker = new CpuTimeTracker(jiffyLengthMs);
+    cGroupsHandler = ResourceHandlerModule.getCGroupsHandler();
+  }
+
+  @Override
+  public long getCumulativeCpuTime() {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#cpu-interface-files
+    return jiffyLengthMs < 0
+        ? UNAVAILABLE
+        : getStat("cpu.stat#usage_usec") * jiffyLengthMs;
+  }
+
+  @Override
+  public long getRssMemorySize(int olderThanAge) {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-interface-files
+    return 1 < olderThanAge
+        ? UNAVAILABLE
+        : getStat("memory.stat#anon");
+  }
+
+  @Override
+  public long getVirtualMemorySize(int olderThanAge) {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-interface-files
+    return 1 < olderThanAge
+        ? UNAVAILABLE
+        : getStat("memory.stat#vmalloc");
+  }
+
+  @Override
+  public void updateProcessTree() {
+    try (Stream<Path> cGroupFiles = Files.list(getCGroupPath())){
+      List<Path> statFiles = cGroupFiles
+          .filter(path -> path.toString().endsWith(".stat"))
+          .collect(Collectors.toList());
+      for (Path statFile : statFiles) {
+        String[] lines = fileToString(statFile).split(System.lineSeparator());
+        for (String line: lines) {
+          String[] parts = line.split(" ");
+          stats.put(statFile.getFileName() + "#" + parts[0], parts[1]);
+        }
+      }
+      cpuTimeTracker.updateElapsedJiffies(
+          BigInteger.valueOf(getStat("cpu.stat#usage_usec")), clock.getTime());
+      LOG.debug("The {} process has the following stat properties updated: {}", pid, stats);
+    } catch (Exception e) {
+      LOG.warn("Failed to read CGroupV2 stats for process: " + pid, e);
+    }
+  }
+
+  private Path getCGroupPath() throws IOException {
+    String relativePath = pid == null
+        ? ResourceHandlerModule.getCGroupsHandler().getRelativePathForCGroup("")
+        : getCGroupRelativePathForPid();
+    return Paths.get(cGroupsHandler.getCGroupMountPath(), relativePath);
+  }
+
+  private String getCGroupRelativePathForPid() throws IOException {
+    // https://docs.kernel.org/admin-guide/cgroup-v2.html#processes
+    String pidCGroupsFile = fileToString(Paths.get(root, "proc", pid, "cgroup"));
```

Comment:
The v1 resource calculator gets the proc fs dir as a parameter. Is there a reason why that couldn't be common thing?

On a quick glance the `CGroupsResourceCalculator.getCGroupRelativePathForPid` method should work here as well, have you maybe tried it?